### DISCUSSION TITLE
copying/deepcopying MeshGeometry objects

### DISCUSSION THREAD

**Discussion by anilbeycorintis** [firedrakeproject/firedrake#3860] [Firedrake support]
Hello,

I am wondering if the `MeshGeometry` objects can be copied as below - or if they are not meant to be copied?

```python
>>> from firedrake import UnitSquareMesh
>>> m1 = UnitSquareMesh(2,2)
>>> from copy import deepcopy
>>> deepcopy(m1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 265, in _reconstruct
    y = func(*args)
        ^^^^^^^^^^^
  File "/usr/lib/python3.11/copyreg.py", line 105, in __newobj__
    return cls.__new__(cls, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: MeshGeometry.__new__() missing 1 required positional argument: 'element'
>>>
```

copy() function gives the same error.

**Comment by dham** [comment]
I can't imagine what the use case for copying that is. If you want a new mesh with the same topology but different geometry, just define a vector lagrange space on the mesh, set its values to whatever coordinates you want and then call `Mesh` on it to get a new mesh.

**Reply by anilbeycorintis** [reply]
Hi, thank you for the quick response!

The example here is a simple reproduction of the issue for clarity. Generally, `MeshGeometry` objects are non-copyable, and a key use-case for copying would be to ensure that modifications to the original don’t affect the duplicate.

Additionally, if a `MeshGeometry` attribute is included within a class or collection, it forces the entire class to lose its copyability, which can be quite limiting.

**Reply by dham** [reply]
That's still a very abstract use case. Do you have a more concrete example in mind?

I think that asking finite element data structures to be generically copyable in the way you are suggesting is not mathematically straightforward. Suppose I have a collection of finite element Functions. If I copy them, what mesh should the result be on? If the original mesh (which is what we do) then the copies are not truly independent because a change to the mesh will affect them all. Alternatively, if you really make deep copies then you now have Functions on a collection of different meshes, which is certainly inefficient and probably not what you mathematically wanted.

The interconnected nature of the data structures (which reflects the mathematical objects that they model) means that deep copying a mesh is always going to be a somewhat bespoke operation, as is the related problem of deep copying Functions.
